<!DOCTYPE html>
<head>

</head>
<body>
    
    <p>fileroute: /home/tjdokas/shops/onr_samples_group/clubhouse/PRS/</p>
    
    <p>inFile: merge</p>
    
    <p>inDir: Phase1/input/</p>
    
    <p>outDir: testout/</p>
    
    <p>geno: 0.01</p>
    
    <p>mind: 0.05</p>
    
    <p>maf: 0.005</p>
    
    <p>indep_pairwise: [50, 5, 0.5]</p>
    
    <p>relatedness: 0.2</p>
    
    <p>1kg_plinkfile: /home/tjdokas/onr_samples_group/GenotypeInfo2015_2018/Files_for_QC/1kG_MDS</p>
    
    <p>1kg_outfile: /treehouse/onr_samples_group/GenotypeInfo2015_2018/Files_for_QC/1kG_MDS_7.bim</p>
    
    
    <img src="testout/Hist-individualMissingness.png">
    <h2>Histogram individual missingness</h2>
    
    <img src="testout/Hist-snpMissingness.png">
    <h2>Histogram SNP missingness</h2>
    
    <img src="testout/Gender_check.png">
    <h2>Gender Sexcheck</h2>
    
    <img src="testout/Men_check.png">
    <h2>Male Sexcheck</h2>
    
    <img src="testout/Women_check.png">
    <h2>Female Sexcheck</h2>
    
    <img src="testout/MAF_distribution.png">
    <h2>MAF Distribution</h2>
    
    <img src="testout/HWE_Histogram.png">
    <h2>Histogram HWE</h2>
    
    <img src="testout/HWE_below_theshold_Histogram.png">
    <h2>Histogram HWE: strongly deviating SNPs only</h2>
    
    <img src="testout/heterozygosity.png">
    <h2>Heterozygosity Rate</h2>
    
    <h2>Initial data information</h2>
    
    <p>individuals info: 257 people (130 males, 127 females) loaded from .fam.</p>
    
    <p>n individuals: 257</p>
    
    <p>SNP info: 706628 variants loaded from .bim file.</p>
    
    <p>n SNP&#39;s: 706628</p>
    
    
    
    <p>On command  --bfile /home/tjdokas/shops/onr_samples_group/clubhouse/PRS/Phase1/input/merge --geno 0.01 --make-bed --out testout/plink: 674076 variants and 257 people pass filters and QC.</p>
    
    
    
    <p>On command  --bfile testout/plink --mind 0.05 --make-bed --out testout/plink: 674076 variants and 255 people pass filters and QC.</p>
    
    
    
    <p>On command  --bfile testout/plink --check-sex --out testout/plink: 674076 variants and 255 people pass filters and QC.</p>
    
    
    
    <p>On command  -bfile testout/plink --remove testout/sex_discrepency.txt --make-bed --out testout/plink: 674076 variants and 249 people pass filters and QC.</p>
    
    
    
    <p>On command  --bfile testout/plink --extract testout/snp_1_22.txt --make-bed --out testout/plink: 656681 variants and 249 people pass filters and QC.</p>
    
    
    
    <p>On command  --bfile testout/plink --maf 0.005 --make-bed --out testout/plink: 616243 variants and 249 people pass filters and QC.</p>
    
    
    
    <p>On command  --bfile testout/plink --hwe 1e-10 --hwe-all --make-bed --out testout/plink: 616110 variants and 249 people pass filters and QC.</p>
    
    
    
    <p>On command  --bfile testout/plink --indep-pairwise 50 5 0.5 --out testout/indepSNP: 616110 variants and 249 people pass filters and QC.</p>
    
    
    
    <p>On command  --bfile testout/plink --extract testout/indepSNP.prune.in --het --out testout/R_hetCheck: 294100 variants and 249 people pass filters and QC.</p>
    
    
    
    <p>On command  --bfile testout/plink --extract testout/indepSNP.prune.in --genome --min 0.2 --out testout/pihat_min0.2: 294100 variants and 249 people pass filters and QC.</p>
    
    
</body>