<!DOCTYPE html>
<head>

</head>
<body>
    <div>
        <ul>
            <li><a href="#Settings">Settings</a></li>
            
            <li><a href="#Histogram individual missingness">Histogram individual missingness</a></li>
            
            <li><a href="#Histogram SNP missingness">Histogram SNP missingness</a></li>
            
            <li><a href="#Gender Sexcheck">Gender Sexcheck</a></li>
            
            <li><a href="#Male Sexcheck">Male Sexcheck</a></li>
            
            <li><a href="#Female Sexcheck">Female Sexcheck</a></li>
            
            <li><a href="#MAF Distribution">MAF Distribution</a></li>
            
            <li><a href="#Histogram HWE">Histogram HWE</a></li>
            
            <li><a href="#Histogram HWE: strongly deviating SNPs only">Histogram HWE: strongly deviating SNPs only</a></li>
            
            <li><a href="#Heterozygosity Rate">Heterozygosity Rate</a></li>
            
            <li><a href="#Parameters">Parameters</a></li>
            <li><a href="#removed">Removed SNPs & Individuals</a></li>
        </ul>
    </div>
    <span id="Settings"><h2>Settings</h2></span>
    
    <p>1kG_dir: /shops/onr_samples_group/clubhouse/GenotypeInfo2015_2018/Files_for_QC/</p>
    
    <p>fileroute: /shops/onr_samples_group/clubhouse/PRS/</p>
    
    <p>inFile: merge</p>
    
    <p>inDir: Phase1/input/</p>
    
    <p>outDir: testout/</p>
    
    <p>geno: 0.01</p>
    
    <p>mind: 0.05</p>
    
    <p>maf: 0.005</p>
    
    <p>indep_pairwise: [50, 5, 0.5]</p>
    
    <p>relatedness: 0.2</p>
    
    
    <img src="testout/Hist-individualMissingness.png">
    <span id="Histogram individual missingness"><h2>Histogram individual missingness</h2></span>
    
    <img src="testout/Hist-snpMissingness.png">
    <span id="Histogram SNP missingness"><h2>Histogram SNP missingness</h2></span>
    
    <img src="testout/Gender_check.png">
    <span id="Gender Sexcheck"><h2>Gender Sexcheck</h2></span>
    
    <img src="testout/Men_check.png">
    <span id="Male Sexcheck"><h2>Male Sexcheck</h2></span>
    
    <img src="testout/Women_check.png">
    <span id="Female Sexcheck"><h2>Female Sexcheck</h2></span>
    
    <img src="testout/MAF_distribution.png">
    <span id="MAF Distribution"><h2>MAF Distribution</h2></span>
    
    <img src="testout/HWE_Histogram.png">
    <span id="Histogram HWE"><h2>Histogram HWE</h2></span>
    
    <img src="testout/HWE_below_theshold_Histogram.png">
    <span id="Histogram HWE: strongly deviating SNPs only"><h2>Histogram HWE: strongly deviating SNPs only</h2></span>
    
    <img src="testout/heterozygosity.png">
    <span id="Heterozygosity Rate"><h2>Heterozygosity Rate</h2></span>
    
    <span id="Parameters"><h2>Parameters</h2></span>
    
    <p>individuals info: 257 people (130 males, 127 females) loaded from .fam.</p>
    
    <p>n individuals: 257</p>
    
    <p>SNP info: 706628 variants loaded from .bim file.</p>
    
    <p>n SNP&#39;s: 706628</p>
    
    <span id="removed"><h2>Removed SNPs & Individuals</h2></span>
    
    
    <p>On command ./bin/plink  --bfile /shops/onr_samples_group/clubhouse/PRS/Phase1/input/merge --geno 0.01 --make-bed --out testout/plink: 674076 variants and 257 people pass filters and QC.</p>
    
    
    
    <p>On command ./bin/plink  --bfile testout/plink --mind 0.05 --make-bed --out testout/plink: 674076 variants and 255 people pass filters and QC.</p>
    
    
    
    <p>On command ./bin/plink  --bfile testout/plink --check-sex --out testout/plink: 674076 variants and 255 people pass filters and QC.</p>
    
    
    
    <p>On command ./bin/plink  -bfile testout/plink --remove testout/sex_discrepency.txt --make-bed --out testout/plink: 674076 variants and 249 people pass filters and QC.</p>
    
    
    
    <p>On command ./bin/plink  --bfile testout/plink --extract testout/snp_1_22.txt --make-bed --out testout/plink: 656681 variants and 249 people pass filters and QC.</p>
    
    
    
    <p>On command ./bin/plink  --bfile testout/plink --maf 0.005 --make-bed --out testout/plink: 616243 variants and 249 people pass filters and QC.</p>
    
    
    
    <p>On command ./bin/plink  --bfile testout/plink --hwe 1e-10 --hwe-all --make-bed --out testout/plink: 616110 variants and 249 people pass filters and QC.</p>
    
    
    
    <p>On command ./bin/plink  --bfile testout/plink --indep-pairwise 50 5 0.5 --out testout/indepSNP: 616110 variants and 249 people pass filters and QC.</p>
    
    
    
    <p>On command ./bin/plink  --bfile testout/plink --extract testout/indepSNP.prune.in --het --out testout/R_hetCheck: 294100 variants and 249 people pass filters and QC.</p>
    
    
    
    <p>On command ./bin/plink  --bfile testout/plink --extract testout/indepSNP.prune.in --genome --min 0.2 --out testout/pihat_min0.2: 294100 variants and 249 people pass filters and QC.</p>
    
    
    
    <p>On command ./bin/plink --bfile /shops/onr_samples_group/clubhouse/GenotypeInfo2015_2018/Files_for_QC/1kG_MDS --set-missing-var-ids @:#[b37]\$1,\$2 --make-bed --out testout/1kG_qc: 25488488 variants and 629 people pass filters and QC.</p>
    
    
    
    <p>On command ./bin/plink --bfile testout/1kG_qc --geno 0.2 --allow-no-sex --make-bed --out testout/1kG_qc: 9007422 variants and 629 people pass filters and QC.</p>
    
    
    
    <p>On command ./bin/plink --bfile testout/1kG_qc --mind 0.2 --allow-no-sex --make-bed --out testout/1kG_qc: 9007422 variants and 629 people pass filters and QC.</p>
    
    
    
    <p>On command ./bin/plink --bfile testout/1kG_qc --geno 0.02 --allow-no-sex --make-bed --out testout/1kG_qc: 8240745 variants and 629 people pass filters and QC.</p>
    
    
    
    <p>On command ./bin/plink --bfile testout/1kG_qc --mind 0.02 --allow-no-sex --make-bed --out testout/1kG_qc: 8240745 variants and 629 people pass filters and QC.</p>
    
    
    
    <p>On command ./bin/plink --bfile testout/1kG_qc --maf 0.05 --allow-no-sex --make-bed --out testout/1kG_qc: 5808310 variants and 629 people pass filters and QC.</p>
    
    
    
    <p>On command ./bin/plink --bfile testout/1kG_qc --hwe 0.001 --allow-no-sex --make-bed --out testout/1kG_qc: 4118334 variants and 629 people pass filters and QC.</p>
    
    
    
    <p>On command ./bin/plink --bfile testout/1kG_qc --extract testout/PopStrat_SNPs.txt --make-bed --out testout/1kG_qc: 401075 variants and 629 people pass filters and QC.</p>
    
    
    
    <p>On command ./bin/plink --bfile testout/plink --extract testout/1kG_MDS_SNPs.txt --recode --make-bed --out testout/PopStrat_MDS: 610062 variants and 249 people pass filters and QC.</p>
    
    
    
    <p>On command ./bin/plink --bfile testout/1kG_qc --update-map testout/buildReport.txt --make-bed --out testout/1kG_qc: 401075 variants and 629 people pass filters and QC.</p>
    
    
    
    <p>On command ./bin/plink --bfile testout/PopStrat_MDS --reference-allele testout/1kG_ref-list.txt --make-bed --out testout/PopStrat-adj: 610062 variants and 249 people pass filters and QC.</p>
    
    
    
    <p>On command ./bin/plink --bfile testout/PopStrat-adj --flip testout/flip_list.txt --reference-allele testout/1kG_ref-list.txt --make-bed --out testout/PopStrat_corrected: 610062 variants and 249 people pass filters and QC.</p>
    
    
    
    <p>On command ./bin/plink --bfile testout/PopStrat_corrected --exclude testout/SNPs_excluded.txt --make-bed --out testout/PopStrat_MDS2: 610017 variants and 249 people pass filters and QC.</p>
    
    
    
    <p>On command ./bin/plink --bfile testout/1kG_qc --exclude testout/SNPs_excluded.txt --make-bed --out testout/1kG_qc: 401058 variants and 629 people pass filters and QC.</p>
    
    
    
    <p>On command ./bin/plink --bfile testout/PopStrat_MDS2 --bmerge testout/1kG_qc.bed testout/1kG_qc.bim testout/1kG_qc.fam --allow-no-sex --make-bed --out testout/MDS_merge: 610017 variants and 878 people pass filters and QC.</p>
    
    
    
    <p>On command ./bin/plink --bfile testout/MDS_merge --extract testout/indepSNP.prune.in --genome --out testout/MDS_merge: 291597 variants and 878 people pass filters and QC.</p>
    
    
    
    <p>On command ./bin/plink --bfile testout/MDS_merge --read-genome testout/MDS_merge.genome --cluster --mds-plot 10 --out testout/MDS_merge: 610017 variants and 878 people pass filters and QC.</p>
    
    
</body>